### PR TITLE
Improvements to ps things

### PR DESCRIPTION
**Pull request by JayH5**
This is mostly in preparation of doing more tests on things like the `ppid` column (the parent PID) so as to assert not just on processes but also the structure of the process tree.

### CODE REVIEW

File: example/docker-helper/docker_helper/utils.py [hunk 1]
```diff
@@ -35,18 +36,39 @@ def output_lines(raw_output, encoding='utf-8'):
...
-    # outside the container in the host's namespaces. Note that this requires
-    # the container to have a 'ps' that responds to the arguments we give it.
-    cols = ','.join(columns)
-    ps_output = container.exec_run(['ps', 'ax', '-o', cols])
-    ps_lines = output_lines(ps_output)
-    ps_lines.pop(0)  # Skip the header
-    ps_entries = [line.split(None, max(0, len(columns) - 1))
-                  for line in ps_lines]
-    # Drop the entry for the ps command itself
-    ps_entries = [e for e in ps_entries if e[2] != 'ps ax -o {}'.format(cols)]
```
Comment (JayH5) [inline example/docker-helper/docker_helper/utils.py:51]: We can't rely on `e[2]`

File: example/docker-helper/docker_helper/utils.py [hunk 2]
```diff
@@ -35,18 +36,39 @@ def output_lines(raw_output, encoding='utf-8'):
...
-    # We use an exec here rather than `container.top()` because we want to run
-    # 'ps' inside the container. This is because we want to get PIDs and
-    # usernames in the container's namespaces. `container.top()` uses 'ps' from
-    # outside the container in the host's namespaces. Note that this requires
-    # the container to have a 'ps' that responds to the arguments we give it.
-    cols = ','.join(columns)
-    ps_output = container.exec_run(['ps', 'ax', '-o', cols])
-    ps_lines = output_lines(ps_output)
-    ps_lines.pop(0)  # Skip the header
-    ps_entries = [line.split(None, max(0, len(columns) - 1))
```
Comment (JayH5) [inline example/docker-helper/docker_helper/utils.py:48]: Not technically correct to split on whitespace as a field could have whitespace in it.

File: example/docker-helper/docker_helper/utils.py [hunk 3]
```diff
@@ -35,18 +36,39 @@ def output_lines(raw_output, encoding='utf-8'):
...
-                  for line in ps_lines]
-    # Drop the entry for the ps command itself
-    ps_entries = [e for e in ps_entries if e[2] != 'ps ax -o {}'.format(cols)]
-    return ps_entries
+PsRow = namedtuple('PsRow', ['pid', 'ruser', 'args'])
```
Comment (JayH5) [inline example/docker-helper/docker_helper/utils.py]: These are just the values we use already. The idea is to add more in future PRs.

File: example/docker-helper/docker_helper/utils.py [hunk 4]
```diff
@@ -35,18 +36,47 @@ def output_lines(raw_output, encoding='utf-8'):
...
+    cmd = ['ps', 'ax', '-o', ','.join(PsRow.columns())]
+    ps_lines = output_lines(container.exec_run(cmd))
+
+    header = ps_lines.pop(0)
+    # Split on the start of the header title words
+    spans = [(match.start(0), match.end(0))
+             for match in re.finditer(r'\b\w+\s*', header)]
+    spans[-1] = (spans[-1][0], None)  # Final span goes to the end of the line
+    ps_entries = [
+        [line[start:end].strip() for start, end in spans] for line in ps_lines]
```
Comment (jerith) [inline example/docker-helper/docker_helper/utils.py:73]: Cunning.

File: example/test.py [hunk 5]
```diff
@@ -55,26 +81,24 @@ def test_expected_processes(self, web_only_container):
 
         assert_that(ps_data, HasLength(5))
```
Comment (jerith) [inline example/test.py]: Do we need this length check? The next two lines pop and assert pid 1, then assert that the remaining entries match what we expect. That should be sufficient, unless there's some kind of issue with `MatchesSetwise` and duplicate entries.

**Review by jerith — COMMENTED** [review]
Looks good.

**Review by jerith — APPROVED** [review]
:+1:

**Pull request merged**